src/Solvers/matrix_iterative_solver.jl

```diff
@@ -0,0 +1,273 @@
+using Oceananigans.Architectures
+using Oceananigans.Architectures: architecture, arch_array
+using Oceananigans.Grids: interior_parent_indices, topology
+using Oceananigans.Fields: interior_copy
+using Oceananigans.Utils: heuristic_workgroup
+using KernelAbstractions: @kernel, @index
+using IterativeSolvers, SparseArrays, LinearAlgebra
+using CUDA, CUDA.CUSPARSE
+
+mutable struct MatrixIterativeSolver{A, G, R, L, D, M, P, S, I, T, F}
+               architecture :: A
```

Comment:
Does it need `architecture`?

Reply:
yeah it is required for the preconditioner